tsinfer/formats.py

```diff
@@ -2174,8 +2185,182 @@ def ancestors_length(self):
+            If None, the information is stored in memory and not persistent.
+        :param int num_flush_threads: See the documentation for :class:`AncestorData`.
+        :param int max_file_size: See the documentation for :class:`AncestorData`.
+
+        :return: A new :class:`.AncestorData` object.
+        :rtype: AncestorData
+        """
+        sample_data._check_finalised()
+        if require_same_sample_data:
+            if sample_data.uuid != self.sample_data_uuid:
+                raise ValueError(
+                    "sample_data differs from that used to build the initial ancestors"
+                )
+        if self.sequence_length != sample_data.sequence_length:
+            raise ValueError("sample_data does not have the correct sequence length")
+
+        if sample_ids is None:
+            sample_ids = []
+            for i in sample_data.individuals():
+                if i.time > 0:
+                    sample_ids += i.samples
+
+        used_sites = np.isin(sample_data.sites_position[:], self.sites_position[:])
+        if np.sum(used_sites) != self.num_sites:
+            raise ValueError(
+                "Some genome positions in the ancestors are not in sample_data"
+            )
+
+        sample_times = np.zeros(len(sample_ids), dtype=self.ancestors_time.dtype)
+        for i, s in enumerate(sample_ids):
+            sample = sample_data.sample(s)
+            if sample.individual != tskit.NULL:
+                sample_times[i] = sample_data.individual(sample.individual).time
+
+        if epsilon is not None:
+            epsilons = np.atleast_1d(epsilon, dtype=self.ancestors_time.dtype)
+            if len(epsilons) == 1:
+                # all get the same epsilon
+                epsilons = np.repeat(epsilons, len(sample_ids))
+        else:
+            anc_times = self.ancestors_time[:][::-1]  # find ascending time order
+            older_index = np.searchsorted(anc_times, sample_times, side="right")
+            # Don't include times older than the oldest ancestor
+            allowed = older_index < self.num_ancestors
+            if np.sum(allowed) > 0:
+                delta_t = anc_times[older_index[allowed]] - sample_times[allowed]
+            else:
+                # All samples have times equal to or older than the oldest curr ancestor
+                time_diffs = np.diff(anc_times)
+                delta_t = np.min(time_diffs[time_diffs > 0])
+            epsilons = np.repeat(np.min(delta_t) / 100.0, len(sample_ids))
+        if len(epsilons) != len(sample_ids):
+            raise ValueError(
+                "The number of epsilon values must equal the number of sample_ids ({})".format(
+                    len(sample_ids)
+                )
+            )
+
+        proxy_times = sample_times + epsilons
+        reverse_time_sorted_indexes = np.argsort(proxy_times)[::-1]
+
+        ancestor_data = self.copy(path=path, max_file_size=max_file_size)
```

Comment:
This seems a bit wasteful since we're writing all the data again in a minute - is the idea to avoid writing out various bits of header information again?

Reply:
It started because I was going to keep the sites_positions array the same, but then persisted b because it means we could keep the same compressors, chunk_sizes, etc. I assumed that a straight like-for-like memory copy would be quick, and we are assured that the new file will have zarr arrays at least as big as the existing ones, so it's not like we can avoid allocating the memory itself. I hoping the resize() operation (in `clear_data()`) wouldn't try to free the zarr memory, but just use what was there previously. But you are bound to know the best way to do this.

Reply:
I think it would be simplest (and easiest to follow) to create a new instance - we're doing a lot of working around and fiddling with internals otherwise. I think the right approach is to pass all the extra details for creating an AncestorData file in as extra kwargs, so we get rid of 
```     
       path=None,
￼        num_flush_threads=0,
￼        max_file_size=None,
```
and replace them with ``**kwargs``, and then use 
```
with AncestorData(**kwargs) as other:
    # build the new file
return other
```
or similar.

Reply:
Done in latest push